tdrs-backend/tdpservice/parsers/case_consistency_validator.py

```diff
@@ -23,6 +24,37 @@ def __init__(self, header, generate_error):
         self.total_cases_cached = 0
         self.total_cases_validated = 0
 
+    def __get_model(self, model_str):
+        """Return a model for the current program type/section given the model's string name."""
+        manager = get_program_model(self.program_type, self.section, model_str)
+        return manager.schemas[0].document.Django.model
+
+    def __generate_and_add_error(self, schema, record, field, msg):
+        """Generate a ParserError and add it to the `generated_errors` list."""
+        err = self.generate_error(
+            error_category=ParserErrorCategoryChoices.CASE_CONSISTENCY,
+            schema=schema,
+            record=record,
+            field=field,
+            error_message=msg,
+        )
+        self.generated_errors.append(err)
+
+    def __get_records_by_rpt_month_year_and_model_type(self):
```

Comment:
This is a big question and might be pointless/moot because it could obscure things.  But I have two thoughts/questions on this. 

First, because we have a partial order over the datafile and a total order of the records in a case, do we need to go through the extra step of creating another data structured to sort them? This seems like duplicated work to me?

Second, if we don't need to do the above, does it make sense to make `record_schema_pairs` an actual class that has separate arrays (or whatever DS makes sense) to keep the T1s, T2s, etc separate so that they could be iterated over or queried for in the same manner that your validators are leveraging the `cases` DS? This way we can update those DS' in that class as we cache the records as opposed to recomputing work that we might be available to us for "free".

Reply:
excellent question! i'd definitely be open to pre-sorting the records as they are inserted into `record_schema_pairs` - i think you're right that it's extra computational work that doesn't necessarily need to be done. 
The reason for creating the sorted object was to have a simple way to look up records that share a `RPT_MONTH_YEAR` (as you point out, the case consistency validator already does the work of sorting by `CASE_NUMBER` which is the other necessary key). To me, it looked moderately cleaner to check `len(t1s)` for a given `RPT_MONTH_YEAR` than to loop over all the records, keeping counts of how many t1s/t2s/t3s etc there are for a `RPT_MONTH_YEAR`. That comes at a computational cost of `O(2n)` or somewhere thereabouts, rather than `O(n)`

Reply:
ended up being a pretty minimal change to get this implemented. i just want to note what we discussed in office hours - the new `SortedRecordSchemaPairs` class maintains two independent copies of the same information. the tradeoff here is memory footprint versus number of iterations; i agree with what you said earlier about the `CaseConsistencyValidator` having a minimal memory footprint in the first place, and making another copy of the ~5-10 cached records wouldn't be a significant increase. having to loop over every record in the datafile again, however, had a noticeable impact on the test run time. thanks again!